torch_xla/csrc/helpers.cpp

```diff
@@ -527,11 +526,69 @@ std::pair<xla::XlaOp, xla::XlaOp> XlaHelpers::PromoteSecondValue(
 
 xla::Shape XlaHelpers::GetPromotedShape(const xla::Shape& shape1,
                                         const xla::Shape& shape2) {
-  return xla::ShapeUtil::MakeShape(
-      shape1.element_type(),
-      torch::lazy::GetPromotedShape(
-          runtime::util::ToVector<int64_t>(shape1.dimensions()),
-          runtime::util::ToVector<int64_t>(shape2.dimensions())));
+  std::vector<int64_t> dimensions;
+  std::vector<bool> dynamic_dimensions;
+
+  // If the rank of a shape is bigger than then other, fill up the first
+  // dimensions with the ones of the bigger.
+  // Example:
+  //   shape1 = [9, 7, 6, 5, 2]
+  //   shape2 =       [6, 1, 2]
+  // Insert [9, 7] into the dimensions vector.
+  if (shape1.dimensions().size() > shape2.dimensions().size()) {
+    dimensions.insert(
+        dimensions.end(), shape1.dimensions().begin(),
+        shape1.dimensions().begin() +
+            (shape1.dimensions().size() - shape2.dimensions().size()));
+    dynamic_dimensions.insert(dynamic_dimensions.end(),
+                              shape1.dynamic_dimensions().begin(),
+                              shape1.dynamic_dimensions().begin() +
+                                  (shape1.dynamic_dimensions().size() -
+                                   shape2.dynamic_dimensions().size()));
+  } else if (shape2.dimensions().size() > shape1.dimensions().size()) {
+    dimensions.insert(
+        dimensions.end(), shape2.dimensions().begin(),
+        shape2.dimensions().begin() +
+            (shape2.dimensions().size() - shape1.dimensions().size()));
+    dynamic_dimensions.insert(dynamic_dimensions.end(),
+                              shape2.dynamic_dimensions().begin(),
+                              shape2.dynamic_dimensions().begin() +
+                                  (shape2.dynamic_dimensions().size() -
+                                   shape1.dynamic_dimensions().size()));
+  }
+  // For the common dimensions, they must match, or one of them be 1.
+  size_t min_size =
+      std::min(shape1.dimensions().size(), shape2.dimensions().size());
+  for (size_t i = 0; i < min_size; i++) {
+    int64_t dim1 =
+        shape1.dimensions()[shape1.dimensions().size() - min_size + i];
+    int64_t dynamic_dim1 =
+        shape1.dynamic_dimensions()[shape1.dynamic_dimensions().size() -
+                                    min_size + i];
+    int64_t dim2 =
+        shape2.dimensions()[shape2.dimensions().size() - min_size + i];
+    int64_t dynamic_dim2 =
+        shape2.dynamic_dimensions()[shape2.dynamic_dimensions().size() -
+                                    min_size + i];
+
+    XLA_CHECK(dim1 == dim2 || dim1 == 1 || dim2 == 1 ||
+              dim1 == xla::Shape::kUnboundedSize ||
+              dim2 == xla::Shape::kUnboundedSize);
+    if (dim1 == 0 || dim2 == 0) {
```

Comment:
Should this be `== 1`? Based on the above check, dim==0 should only be possible if they're both 0?

Reply:
There could be other cases like `1,0` or `0,1` which would also make the above check hold good.

Reply:
Can we rewrite this to use the following algorithm:

1. Pad shapes with 1 using `ExtractDimensionSizesWithPadding` so ranks match.
2. Call `[xla::InferDegenerateDimensionBroadcastShape](https://github.com/openxla/xla/blame/b03bcf70e0de37eb532b19aef9953a91cb4b3e0b/xla/service/shape_inference.cc#L943)` to compute final shape.

For (2), this method is private currently, we can refactor. But in the meantime we should be able to use `[InferBinaryOpShape](https://github.com/openxla/xla/blob/b03bcf70e0de37eb532b19aef9953a91cb4b3e0b/xla/service/shape_inference.cc#L1197)` with kAdd while we refactor (I think we're using elementwise semantics here?)

Reply:
That is a good point! And we should aim for it. That way PyTorch/XLA code-base can have a sync between the broadcasting logic with XLA.

However, other than the refactoring that you pointed out we might need additional refactoring. For example:
[InferBinaryOpShape](https://github.com/openxla/xla/blob/b03bcf70e0de37eb532b19aef9953a91cb4b3e0b/xla/service/shape_inference.cc#L1197) calls [InferElementwiseBinaryOpShape](https://github.com/openxla/xla/blob/b03bcf70e0de37eb532b19aef9953a91cb4b3e0b/xla/service/shape_inference.cc#L1134) which has [check](https://github.com/openxla/xla/blob/b03bcf70e0de37eb532b19aef9953a91cb4b3e0b/xla/service/shape_inference.cc#L1140) for same element type. 

test/test_core_aten_ops.py has tests like [AtenOpTest::test_aten_add_Scalar_2](https://github.com/pytorch/xla/blob/9c320743396ed6ffc50d37892cb07bdde1012b59/test/test_core_aten_ops.py#L315) which can generate shapes like ` s32[10,10] and f32[10,10]` which will fail at that point.

Instead we do the following before we making the XLA shape inference APIs usable in the current context:

1. Refactor `getPromotedShape` to match the broadcasting logic in [xla::InferDegenerateDimensionBroadcastShape](https://github.com/openxla/xla/blame/b03bcf70e0de37eb532b19aef9953a91cb4b3e0b/xla/service/shape_inference.cc#L943).
2. We do not have to pad with 1's before calling [xla::InferDegenerateDimensionBroadcastShape](https://github.com/openxla/xla/blame/b03bcf70e0de37eb532b19aef9953a91cb4b3e0b/xla/service/shape_inference.cc#L943). Instead we can follow the same logic  used in [ShapeInference::InferInDimBroadcastShape](https://github.com/openxla/xla/blob/b03bcf70e0de37eb532b19aef9953a91cb4b3e0b/xla/service/shape_inference.cc#L1043), which is pad the smaller ranked shape with ones larger ones. This is what we already proposed at the beginning.
3. Have a ticket to track the refactoring needed in XLA level.

With (1) it would be a matter of time to replace the broadcasting logic with call to intended API. There will be no change in the exported code.

Reply:
Sounds like a plan! Let's have a ticket to track some of these cleanup tasks so we don't forget, otherwise sounds great! Feel free to resolve this thread.